File: home-modules/tools/i3_project_monitor/daemon_client.py
"""JSON-RPC client for daemon communication.

Handles connection, reconnection, and JSON-RPC requests to the i3 project daemon.
"""

import asyncio
import json
import logging
import os
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, Optional, AsyncIterator, List

from i3ipc import aio as i3ipc_aio

from .models import ConnectionState, OutputState, WorkspaceAssignment

logger = logging.getLogger(__name__)


class DaemonClient:
    """JSON-RPC client for i3 project daemon communication."""

    def __init__(self, socket_path: Optional[str] = None):
        """Initialize daemon client.

        Args:
            socket_path: Path to daemon Unix socket (default: systemd runtime dir)
        """
        if socket_path is None:
            runtime_dir = os.environ.get("XDG_RUNTIME_DIR", f"/run/user/{os.getuid()}")
            socket_path = f"{runtime_dir}/i3-project-daemon/ipc.sock"

        self.state = ConnectionState(socket_path=socket_path)
        self.reader: Optional[asyncio.StreamReader] = None
        self.writer: Optional[asyncio.StreamWriter] = None
        self.request_id: int = 0

        # i3 IPC connection for direct queries (Feature 018)
        self.i3: Optional[i3ipc_aio.Connection] = None

    async def connect(self) -> None:
        """Connect to daemon socket.

        Raises:
            ConnectionError: If connection fails
        """
        try:
            self.reader, self.writer = await asyncio.open_unix_connection(self.state.socket_path)
            self.state.connected = True
            self.state.last_success_at = datetime.now()
            self.state.connection_attempts = 0
            logger.info(f"Connected to daemon at {self.state.socket_path}")

        except Exception as e:
            self.state.connected = False
            self.state.last_error = str(e)
            self.state.last_failure_at = datetime.now()
            self.state.connection_attempts += 1
            logger.error(f"Failed to connect to daemon: {e}")
            raise ConnectionError(f"Failed to connect to daemon: {e}")

    async def connect_with_retry(self) -> None:
        """Connect to daemon with exponential backoff retry.

        Raises:
            ConnectionError: If all retry attempts fail
        """
        while self.state.should_retry():
            try:
                await self.connect()
                return  # Success
            except ConnectionError:
                if not self.state.should_retry():
                    raise

                delay = self.state.get_retry_delay()
                logger.warning(f"Retry {self.state.connection_attempts}/{self.state.max_retries} in {delay}s...")
                await asyncio.sleep(delay)

        raise ConnectionError(f"Failed to connect after {self.state.max_retries} attempts")

    async def disconnect(self) -> None:
        """Close connection to daemon."""
        if self.writer:
            self.writer.close()
            await self.writer.wait_closed()
            self.writer = None
            self.reader = None

        self.state.connected = False
        logger.info("Disconnected from daemon")

    async def request(self, method: str, params: Optional[Dict[str, Any]] = None) -> Any:
        """Send JSON-RPC request and wait for response.

        Args:
            method: JSON-RPC method name
            params: Optional method parameters

        Returns:
            JSON-RPC result field

        Raises:
            ConnectionError: If not connected or connection lost
            ValueError: If daemon returns error response
        """
        if not self.state.connected or not self.writer or not self.reader:
            raise ConnectionError("Not connected to daemon")

        self.request_id += 1
        request = {
            "jsonrpc": "2.0",
            "method": method,
            "params": params or {},
            "id": self.request_id
        }

        try:
            # Send request
            message = json.dumps(request).encode() + b"\n"
            self.writer.write(message)
            await self.writer.drain()

            # Read response
            data = await self.reader.readline()
            if not data:
                raise ConnectionError("Connection closed by daemon")

            response = json.loads(data.decode())

            # Check for error
            if "error" in response:
                error = response["error"]
                raise ValueError(f"Daemon error {error['code']}: {error['message']}")

            return response.get("result")

        except (ConnectionError, BrokenPipeError, asyncio.IncompleteReadError) as e:
            self.state.connected = False
            self.state.last_error = str(e)
            logger.error(f"Connection lost during request: {e}")
            raise ConnectionError(f"Connection lost: {e}")

    async def get_status(self) -> Dict[str, Any]:
        """Query daemon status.

        Returns:
            Status dictionary from daemon
        """
        return await self.request("get_status")

    async def get_active_project(self) -> Dict[str, Any]:
        """Get active project info.

        Returns:
            Active project dictionary
        """
        return await self.request("get_active_project")

    async def get_projects(self) -> Dict[str, Any]:
        """List all projects.

        Returns:
            Projects dictionary
        """
        return await self.request("get_projects")

    async def get_windows(self, project: Optional[str] = None) -> Dict[str, Any]:
        """Query windows with optional project filter.

        Args:
            project: Optional project name filter

        Returns:
            Windows dictionary
        """
        params = {"project": project} if project else {}
        return await self.request("get_windows", params)

    async def get_events(self, limit: int = 100, event_type: Optional[str] = None) -> Dict[str, Any]:
        """Query recent events.

        Args:
            limit: Maximum number of events to return
            event_type: Optional event type filter (e.g., "window", "workspace")

        Returns:
            Events dictionary with events list and stats
        """
        params = {"limit": limit}
        if event_type:
            params["event_type"] = event_type

        return await self.request("get_events", params)

    async def list_monitors(self) -> Dict[str, Any]:
        """List connected monitor clients.

        Returns:
            Monitors dictionary
        """
        return await self.request("list_monitors")

    async def subscribe_events(self, event_types: Optional[list[str]] = None) -> Dict[str, Any]:
        """Subscribe to event stream.

        Args:
            event_types: Optional list of event types to subscribe to (empty = all)

        Returns:
            Subscription confirmation
        """
        params = {"subscribe": True}
        if event_types:
            params["event_types"] = event_types

        return await self.request("subscribe_events", params)

    async def unsubscribe_events(self) -> Dict[str, Any]:
        """Unsubscribe from event stream.

        Returns:
            Unsubscribe confirmation
        """
        return await self.request("subscribe_events", {"subscribe": False})

    async def stream_events(self) -> AsyncIterator[Dict[str, Any]]:
        """Stream events from daemon after subscribing.

        Yields:
            Event notification dictionaries

        Raises:
            ConnectionError: If not connected
        """
        if not self.state.connected or not self.reader:
            raise ConnectionError("Not connected to daemon")

        logger.info("Starting event stream")

        while True:
            try:
                data = await self.reader.readline()
                if not data:
                    logger.warning("Connection closed by daemon")
                    break

                message = json.loads(data.decode())

                # Handle JSON-RPC notification (no id field)
                if "method" in message and message.get("method") == "event_notification":
                    yield message.get("params", {})

            except (json.JSONDecodeError, asyncio.IncompleteReadError) as e:
                logger.error(f"Error reading event stream: {e}")
                break
            except Exception as e:
                logger.error(f"Unexpected error in event stream: {e}")
                break

        self.state.connected = False
        logger.info("Event stream ended")

    async def connect_i3(self) -> None:
        """Connect to i3 IPC for direct queries (Feature 018).

        This creates a separate i3 IPC connection for querying i3 state directly
        (GET_OUTPUTS, GET_WORKSPACES, GET_TREE) without going through the daemon.

        Raises:
            ConnectionError: If i3 connection fails
        """
        try:
            self.i3 = await i3ipc_aio.Connection().connect()
            logger.info("Connected to i3 IPC")
        except Exception as e:
            logger.error(f"Failed to connect to i3 IPC: {e}")
            raise ConnectionError(f"Failed to connect to i3 IPC: {e}")

    async def disconnect_i3(self) -> None:
        """Disconnect from i3 IPC."""
        if self.i3:
            # i3ipc.aio doesn't have an explicit disconnect, connections are auto-managed
            self.i3 = None
            logger.info("Disconnected from i3 IPC")

    async def get_i3_outputs(self) -> List[OutputState]:
        """Query i3 outputs using GET_OUTPUTS IPC (Feature 018).

        Returns:
            List of OutputState objects from i3

        Raises:
            ConnectionError: If not connected to i3
        """
        if not self.i3:
            await self.connect_i3()

        try:
            outputs = await self.i3.get_outputs()
            return [OutputState.from_i3_output(output) for output in outputs]
        except Exception as e:
            logger.error(f"Failed to get i3 outputs: {e}")
            raise ConnectionError(f"Failed to get i3 outputs: {e}")

    async def get_i3_workspaces(self) -> List[WorkspaceAssignment]:
        """Query i3 workspaces using GET_WORKSPACES IPC (Feature 018).

        Returns:
            List of WorkspaceAssignment objects from i3

        Raises:
            ConnectionError: If not connected to i3
        """
        if not self.i3:
            await self.connect_i3()

        try:
            workspaces = await self.i3.get_workspaces()
            return [WorkspaceAssignment.from_i3_workspace(ws) for ws in workspaces]
        except Exception as e:
            logger.error(f"Failed to get i3 workspaces: {e}")
            raise ConnectionError(f"Failed to get i3 workspaces: {e}")

File: home-modules/tools/i3pm/src/services/daemon-client.ts
/**
 * Daemon client service
 * Feature 035: Registry-Centric Project & Workspace Management
 *
 * JSON-RPC 2.0 client for communicating with the Python i3pm daemon
 */

export class DaemonError extends Error {
  constructor(message: string, public code?: number, public override cause?: Error) {
    super(message);
    this.name = "DaemonError";
  }
}

function isIgnorableSocketCloseError(error: unknown): boolean {
  if (!(error instanceof Error)) {
    return false;
  }
  const message = String(error.message || "");
  return message.includes("Bad resource ID")
    || message.includes("Broken pipe")
    || message.includes("not connected")
    || message.includes("closed");
}

/**
 * JSON-RPC 2.0 request
 */
interface JsonRpcRequest {
  jsonrpc: "2.0";
  method: string;
  params?: unknown;
  id: number | string;
}

/**
 * JSON-RPC 2.0 response (success)
 */
interface JsonRpcSuccess {
  jsonrpc: "2.0";
  result: unknown;
  id: number | string;
}

/**
 * JSON-RPC 2.0 response (error)
 */
interface JsonRpcError {
  jsonrpc: "2.0";
  error: {
    code: number;
    message: string;
    data?: unknown;
  };
  id: number | string;
}

type JsonRpcResponse = JsonRpcSuccess | JsonRpcError;

/**
 * Check if response is an error
 */
function isJsonRpcError(response: JsonRpcResponse): response is JsonRpcError {
  return "error" in response;
}

/**
 * Daemon client for JSON-RPC communication
 */
export class DaemonClient {
  private socketPath: string;
  private requestId = 0;
  private conn: Deno.UnixConn | null = null;

  constructor(socketPath?: string) {
    // Feature 117: User socket only (daemon runs as user service)
    if (socketPath) {
      this.socketPath = socketPath;
    } else {
      const runtimeDir = Deno.env.get("XDG_RUNTIME_DIR") ?? `/run/user/${Deno.uid()}`;
      this.socketPath = `${runtimeDir}/i3-project-daemon/ipc.sock`;
    }
  }

  /**
   * Connect to daemon socket
   */
  async connect(): Promise<void> {
    if (this.conn) {
      return; // Already connected
    }

    try {
      this.conn = await Deno.connect({ path: this.socketPath, transport: "unix" });
    } catch (error) {
      throw new DaemonError(
        `Failed to connect to daemon at ${this.socketPath}. Is the daemon running?`,
        -1,
        error instanceof Error ? error : undefined,
      );
    }
  }

  /**
   * Disconnect from daemon
   */
  disconnect(): void {
    if (this.conn) {
      try {
        this.conn.close();
      } catch (error) {
        if (!isIgnorableSocketCloseError(error)) {
          throw error;
        }
      }
      this.conn = null;
    }
  }

  /**
   * Compatibility alias for callers that expect an async close method.
   */
  async close(): Promise<void> {
    this.disconnect();
  }

  /**
   * Send JSON-RPC request and wait for response
   */
  async request<T = unknown>(method: string, params?: unknown): Promise<T> {
    await this.connect();

    if (!this.conn) {
      throw new DaemonError("Not connected to daemon");
    }

    const id = ++this.requestId;
    const request: JsonRpcRequest = {
      jsonrpc: "2.0",
      method,
      params,
      id,
    };

    // Send request
    const requestData = JSON.stringify(request) + "\n";
    const encoder = new TextEncoder();
    await this.conn.write(encoder.encode(requestData));

    // Read one newline-delimited JSON response. Large dashboard payloads routinely
    // exceed 64KB, so a single fixed-size read is not reliable.
    const decoder = new TextDecoder();
    const buffer = new Uint8Array(16384);
    const TIMEOUT_MS = 30000;
    let responseText = "";

    while (true) {
      const readPromise = this.conn.read(buffer);
      let timeoutHandle: number | null = null;
      const timeoutPromise = new Promise<never>((_, reject) => {
        timeoutHandle = setTimeout(
          () => reject(new DaemonError(`Daemon response timeout after ${TIMEOUT_MS}ms`)),
          TIMEOUT_MS,
        );
      });

      let bytesRead: number | null;
      try {
        bytesRead = await Promise.race([readPromise, timeoutPromise]) as number | null;
      } catch (e) {
        await this.close();
        throw e;
      } finally {
        if (timeoutHandle !== null) {
          clearTimeout(timeoutHandle);
        }
      }

      if (bytesRead === null) {
        throw new DaemonError("Connection closed by daemon");
      }

      responseText += decoder.decode(buffer.subarray(0, bytesRead));
      if (responseText.includes("\n")) {
        responseText = responseText.split("\n", 1)[0];
        break;
      }
    }

    const response: JsonRpcResponse = JSON.parse(responseText);

    if (isJsonRpcError(response)) {
      throw new DaemonError(response.error.message, response.error.code);
    }

    return response.result as T;
  }

  /**
   * Send JSON-RPC request (alias for request)
   * Used by monitors.ts and other commands for compatibility
   */
  async sendRequest<T = unknown>(method: string, params?: unknown): Promise<T> {
    return await this.request<T>(method, params);
  }

  /**
   * Get daemon status
   */
  async getStatus(): Promise<{
    status: string;
    connected: boolean;
    uptime: number;
    active_project: string | null;
    window_count: number;
    workspace_count: number;
    event_count: number;
    error_count: number;
    version: string;
    socket_path: string;
  }> {
    return await this.request("get_status");
  }

  /**
   * Get window state (i3 tree query)
   */
  async getWindowState(params?: {
    include_geometry?: boolean;
    workspace_filter?: number[];
  }): Promise<unknown> {
    return await this.request("get_window_state", params);
  }

  /**
   * Get active project from daemon
   */
  async getActiveProject(): Promise<string | null> {
    const status = await this.getStatus();
    return status.active_project;
  }

  async getActiveContext<T = unknown>(): Promise<T> {
    return await this.request<T>("context.current", {});
  }

  async ensureContext<T = unknown>(params: {
    qualified_name?: string;
    project_name?: string;
    target_variant?: "local" | "ssh";
    prefer_local?: boolean;
    clear?: boolean;
  }): Promise<T> {
    return await this.request<T>("context.ensure", params);
  }

  async getRuntimeSnapshot<T = unknown>(): Promise<T> {
    return await this.request<T>("dashboard.snapshot", {});
  }

  async getDashboardSnapshot<T = unknown>(): Promise<T> {
    return await this.request<T>("dashboard.snapshot", {});
  }

  /**
   * Close project windows (used before layout restore)
   */
  async closeProjectWindows(projectName: string): Promise<number> {
    const result = await this.request<{ closed_count: number }>(
      "close_project_windows",
      { project_name: projectName },
    );
    return result.closed_count;
  }

  /**
   * Send project switch notification to daemon
   * Daemon will filter windows based on /proc reading
   */
  async notifyProjectSwitch(projectName: string): Promise<void> {
    await this.request("notify_project_switch", { project_name: projectName });
  }

  /**
   * Clear project (return to global mode)
   */
  async notifyProjectClear(): Promise<void> {
    await this.request("notify_project_clear");
  }

  /**
   * Get recent daemon events (for debugging)
   */
  async getEvents(params?: {
    limit?: number;
    event_type?: string;
  }): Promise<unknown[]> {
    return await this.request("get_events", params);
  }

  /**
   * Subscribe to daemon event stream (async iterator)
   * Yields events as they are broadcast by the daemon
   */
  async *subscribeToEvents(): AsyncIterableIterator<{ type: string; data: unknown; timestamp: string }> {
    if (!this.conn) {
      throw new DaemonError("Not connected to daemon");
    }

    // Subscribe to events
    await this.request("subscribe_events", { subscribe: true });

    // Create text decoder for reading lines
    const decoder = new TextDecoder();
    let partialLine = "";

    try {
      // Read incoming messages (event broadcasts)
      for await (const chunk of this.conn.readable) {
        // Decode chunk and split by newlines
        const text = partialLine + decoder.decode(chunk);
        const lines = text.split("\n");

        // Keep the last incomplete line
        partialLine = lines.pop() || "";

        // Process complete lines
        for (const line of lines) {
          if (!line.trim()) continue;

          try {
            const message = JSON.parse(line);

            // Daemon broadcasts events as JSON-RPC notifications:
            // { jsonrpc: "2.0", method: "event_notification", params: {...} }
            if (message.method === "event_notification" && message.params) {
              const eventData = message.params;
              yield {
                type: eventData.event_type || "unknown",
                data: eventData,
                timestamp: eventData.timestamp || new Date().toISOString(),
              };
            }
          } catch (error) {
            // Ignore parse errors for now (might be fragmented JSON)
            continue;
          }
        }
      }
    } finally {
      // Unsubscribe when done
      try {
        await this.request("subscribe_events", { subscribe: false });
      } catch {
        // Ignore errors during cleanup
      }
    }
  }

  /**
   * Subscribe to daemon state change notifications.
   * Yields lightweight invalidation events that consumers can use to refetch
   * heavier dashboard/session state only when needed.
   */
  async *subscribeToStateChanges(): AsyncIterableIterator<{
    type: string;
    timestamp: number;
    snapshot_version?: number;
    session_generation?: number;
    display_generation?: number;
  }> {
    if (!this.conn) {
      throw new DaemonError("Not connected to daemon");
    }

    await this.request("subscribe_state_changes", {});

    const decoder = new TextDecoder();
    let partialLine = "";

    for await (const chunk of this.conn.readable) {
      const text = partialLine + decoder.decode(chunk);
      const lines = text.split("\n");
      partialLine = lines.pop() || "";

      for (const line of lines) {
        if (!line.trim()) continue;

        try {
          const message = JSON.parse(line);
          if (message.method !== "state_changed" || !message.params) {
            continue;
          }

          const params = message.params as Record<string, unknown>;
          yield {
            type: String(params.type || "state_changed"),
            timestamp: Number(params.timestamp || Date.now()),
            snapshot_version: params.snapshot_version === undefined ? undefined : Number(params.snapshot_version),
            session_generation: params.session_generation === undefined ? undefined : Number(params.session_generation),
            display_generation: params.display_generation === undefined ? undefined : Number(params.display_generation),
          };
        } catch {
          // Ignore malformed notification lines and continue consuming the stream.
        }
      }
    }
  }

  /**
   * Ping daemon (health check)
   */
  async ping(): Promise<boolean> {
    try {
      const status = await this.getStatus();
      return status.status === "running";
    } catch {
      return false;
    }
  }

  /**
   * Get launch registry statistics (Feature 041)
   *
   * Returns launch notification and correlation metrics from the daemon.
   */
  async getLaunchStats(): Promise<{
    total_pending: number;
    unmatched_pending: number;
    total_notifications: number;
    total_matched: number;
    total_expired: number;
    total_failed_correlation: number;
    match_rate: number;
    expiration_rate: number;
  }> {
    return await this.request("get_launch_stats");
  }

  /**
   * Get pending launches (Feature 041)
   *
   * Debug endpoint to inspect currently pending launch notifications.
   *
   * @param includeMatched - Include already-matched launches (default: false)
   */
  async getPendingLaunches(includeMatched = false): Promise<{
    pending_launches: Array<{
      launch_id: string;
      app_name: string;
      project_name: string;
      expected_class: string;
      workspace_number: number;
      age: number;
      matched: boolean;
    }>;
    count: number;
  }> {
    return await this.request("get_pending_launches", { include_matched: includeMatched });
  }

  /**
   * Get applications from daemon's registry
   *
   * Query the daemon's in-memory application registry to see what apps
   * it has loaded. Useful for debugging configuration sync issues.
   *
   * @param params - Optional filter parameters
   * @param params.name - Filter by application name
   * @param params.scope - Filter by scope (scoped/global)
   * @param params.workspace - Filter by preferred workspace number
   */
  async getDaemonApps(params?: {
    name?: string;
    scope?: string;
    workspace?: number;
  }): Promise<{
    applications: Array<unknown>;
    version: string;
    count: number;
    registry_path: string;
  }> {
    return await this.request("daemon.apps", params);
  }

  /**
   * Get daemon socket path
   */
  getSocketPath(): string {
    return this.socketPath;
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.conn !== null;
  }
}

/**
 * Helper function to execute daemon request with automatic connection management
 */
export async function withDaemonClient<T>(
  fn: (client: DaemonClient) => Promise<T>,
  socketPath?: string,
): Promise<T> {
  const client = new DaemonClient(socketPath);
  try {
    await client.connect();
    return await fn(client);
  } finally {
    client.disconnect();
  }
}
